src/Microsoft.DotNet.Build.Tasks/PackageFiles/versioning.targets

```diff
@@ -151,6 +154,75 @@
+      <CurrentVersionLines Include="%3CPropertyGroup%3E" />
+      <CurrentVersionLines Include="%3CBuildNumberMajor Condition=%22%27%24(BuildNumberMajor)%27==%27%27%22%3E$(OverrideBuildNumberMajor)%3C/BuildNumberMajor%3E" />
+      <CurrentVersionLines Include="%3C/PropertyGroup%3E" />
+      <CurrentVersionLines Include="%3C/Project%3E" />
+      <BuildVersionFileItem Include="$(BuildVersionFile)" />
+    </ItemGroup>
+
+    <PropertyGroup>
+      <BuildVersionSemaphoreFile>%(BuildVersionFileItem.RootDir)%(BuildVersionFileItem.Directory)semaphore.txt</BuildVersionSemaphoreFile>
+      <ShouldLockFileWrite Condition="!Exists('$(BuildVersionSemaphoreFile)')">true</ShouldLockFileWrite>
+    </PropertyGroup>
+
+    <MakeDir Condition="!Exists('%(BuildVersionFileItem.RootDir)%(BuildVersionFileItem.Directory)')" Directories="%(BuildVersionFileItem.RootDir)%(BuildVersionFileItem.Directory)" />
+
+    <Touch AlwaysCreate="true" Files="$(BuildVersionSemaphoreFile)" Condition="'$(ShouldLockFileWrite)'=='true'" />
+
+    <WriteLinesToFile
+      ContinueOnError="WarnAndContinue"
+      Condition="'$(ShouldLockFileWrite)'=='true'"
+      File="$(BuildVersionFile)"
+      Lines="@(CurrentVersionLines)"
+      Overwrite="true" />
+
+  </Target>
+
+  <Target Name="GenerateCurrentVersion" DependsOnTargets="GetLatestCommitDate">
+    <Error Condition="'$(DateToUse)'==''" Text="A date is needed to Generate the current version." />
+
+    <GenerateCurrentVersion SeedDate="$(DateToUse)">
+      <Output PropertyName="OverrideBuildNumberMajor" TaskParameter="GeneratedVersion" />
+    </GenerateCurrentVersion>
+  </Target>
+
+  <Target Name="GetLatestCommitDate" Condition="'$(DateToUse)'==''">
+    <Exec Command="git show -s --format=^%25%25cd --date=short HEAD" IgnoreExitCode="true" IgnoreStandardErrorWarningFormat="true" ConsoleToMSBuild="true">
+      <Output TaskParameter="ConsoleOutput" PropertyName="DateToUse" />
+    </Exec>
+  </Target>
+
+  <Target Name="GetTargetsFileLastModifiedDate" Condition="'$(DateToUse)'==''" AfterTargets="GetLatestCommitDate">
+    <ItemGroup>
+      <VersionTargetsFile Include="$(MSBuildThisFileFullPath)" />
+    </ItemGroup>
+
+    <PropertyGroup>
+      <DateToUse>%(VersionTargetsFile.ModifiedTime)</DateToUse>
+    </PropertyGroup>
+  </Target>
+
+  <!-- This target will only be executed if BuildValues.props doesn't exist yet -->
+  <Target Name="CreateVersionFileDuringBuild" Condition="'$(SkipVersionGeneration)'!='true' AND '$(ShouldCreateVersionFileDuringBuild)'=='true'" DependsOnTargets="CreateOrUpdateCurrentVersionFile">
+    <PropertyGroup>
+      <BuildNumberMajor>$(OverrideBuildNumberMajor)</BuildNumberMajor>
+      <AssemblyFileVersion>$(MajorVersion).$(MinorVersion).$(BuildNumberMajor).$(BuildNumberMinor)</AssemblyFileVersion>
+      <VersionSuffix Condition="'$(PreReleaseLabel)' != ''">-$(PreReleaseLabel)</VersionSuffix>
```

Comment:
Yes, We basically need to reconstruct the versionsuffix since by the time that this target executes, the VersionSuffix is already defined, so we need to override it with the new BuildNumberMajor we have calculated.